### PR TITLE
Move Snapshot APIs in-tree

### PR DESCRIPTION
**Pull request by xing-yang**
**What this PR does / why we need it**:
This PR moves snapshot APIs in-tree and also adds a snapshot controller in-tree.
The design proposal is here: kubernetes/community#1662

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #
https://github.com/kubernetes/features/issues/543

**Special notes for your reviewer**:

**Release note**:

```release-note
Volume snapshot APIs are moved in-tree.  Snapshot feature is added for in-tree and out-of-tree CSI plugins. 
```

### CODE REVIEW

**Comment by dims** [comment]
@kubernetes/sig-storage-api-reviews 
@kubernetes/api-reviewers 

/sig storage

**Comment by dixudx** [comment]
I'd prefer to keep this PR and close #64365. Not need to split a single feature into two separate PRs, which will be inconvenient for integration tests.

**Comment by wackxu** [comment]
cc

File: pkg/apis/storage/types.go [hunk 1]
```diff
@@ -80,6 +80,12 @@ type StorageClass struct {
 	// the DynamicProvisioningScheduling feature.
 	// +optional
 	AllowedTopologies []api.TopologySelectorTerm
+
+	// Snapshotter is the driver expected to handle volume snapshotting.
+	// This is an optionally-prefixed name, like a label key.
+	// For example: "kubernetes.io/gce-pd" or "kubernetes.io/aws-ebs".
+	// This value may not be empty.
+	Snapshotter string
```
Comment (jingxu97) [inline pkg/apis/storage/types.go]: We may not need this if provisioner is sufficient for this purpose.

File: pkg/controller/volume/snapshot/snapshot_controller.go [hunk 2]
```diff
@@ -0,0 +1,1470 @@
...
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+	"k8s.io/kubernetes/pkg/controller"
+	"k8s.io/kubernetes/pkg/controller/volume/events"
+	"k8s.io/kubernetes/pkg/util/goroutinemap"
+	"k8s.io/kubernetes/pkg/util/goroutinemap/exponentialbackoff"
+	vol "k8s.io/kubernetes/pkg/volume"
+)
+
+// Number of retries when we create a VSD object for a created snapshot.
```
Comment (jingxu97) [inline pkg/controller/volume/snapshot/snapshot_controller.go]: VSD --> volume snapshot data or VolumeSnapshotData.

**Comment by rootfs** [comment]
@xing-yang need a release note

File: pkg/controller/volume/snapshot/snapshot_controller.go [hunk 3]
```diff
@@ -0,0 +1,1470 @@
...
+		createSnapshotDataRetryCount: createSnapshotDataRetryCount,
+		createSnapshotDataInterval:   createSnapshotDataInterval,
+		runningOperations:            goroutinemap.NewGoRoutineMap(true),
+		snapshotStore:                cache.NewStore(cache.DeletionHandlingMetaNamespaceKeyFunc),
+		snapshotDataStore:            cache.NewStore(cache.DeletionHandlingMetaNamespaceKeyFunc),
+		snapshotQueue:                workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "csi-snapshotter-vs"),
+		snapshotDataQueue:            workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "csi-snapshotter-vsd"),
+	}
+
+	// Prober is nil because vsd is not aware of Flexvolume.
```
Comment (jingxu97) [inline pkg/controller/volume/snapshot/snapshot_controller.go]: vsd --> volume snapshot

File: pkg/controller/volume/snapshot/snapshot_controller.go [hunk 4]
```diff
@@ -0,0 +1,1470 @@
...
+	ctrl.volumeLister = p.VolumeInformer.Lister()
+	ctrl.volumeListerSynced = p.VolumeInformer.Informer().HasSynced
+
+	ctrl.claimLister = p.ClaimInformer.Lister()
+	ctrl.claimListerSynced = p.ClaimInformer.Informer().HasSynced
+
+	return ctrl, nil
+}
+
+// enqueueWork adds vs or vsd to given work queue.
```
Comment (jingxu97) [inline pkg/controller/volume/snapshot/snapshot_controller.go]: vs -- >volume snapshot

File: pkg/controller/volume/snapshot/snapshot_controller.go [hunk 5]
```diff
@@ -0,0 +1,1470 @@
...
+	ClusterName                string
+	VolumeInformer             coreinformers.PersistentVolumeInformer
+	ClaimInformer              coreinformers.PersistentVolumeClaimInformer
+	VolumeSnapshotInformer     storageinformers.VolumeSnapshotInformer
+	VolumeSnapshotDataInformer storageinformers.VolumeSnapshotDataInformer
+	EventRecorder              record.EventRecorder
+}
+
+type SnapshotController struct {
+	vsLister           storagelisters.VolumeSnapshotLister
```
Comment (jingxu97) [inline pkg/controller/volume/snapshot/snapshot_controller.go]: maybe snapshotLister and snapshotDataLister?

File: pkg/controller/volume/snapshot/snapshot_controller.go (hunk 2 above)
Comment (xing-yang) [reply pkg/controller/volume/snapshot/snapshot_controller.go]: Sure.

File: pkg/controller/volume/snapshot/snapshot_controller.go (hunk 5 above)
Comment (xing-yang) [reply pkg/controller/volume/snapshot/snapshot_controller.go]: VolumeSnapshotLister is an auto-generated name.  The only way to change it is to rename the API object from VolumeSnapshot to Snapshot and then re-generate the Lister.

File: pkg/controller/volume/snapshot/snapshot_controller.go (hunk 3 above)
Comment (xing-yang) [reply pkg/controller/volume/snapshot/snapshot_controller.go]: Sure.

File: pkg/controller/volume/snapshot/snapshot_controller.go (hunk 4 above)
Comment (xing-yang) [reply pkg/controller/volume/snapshot/snapshot_controller.go]: Sure.

File: pkg/controller/volume/snapshot/snapshot_controller.go (hunk 5 above)
Comment (xing-yang) [reply pkg/controller/volume/snapshot/snapshot_controller.go]: @jingxu97 You meant to rename "vsLister" to "snapshotLister"?  Sure.

**Comment by xing-yang** [comment]
The PR is moved to https://github.com/kubernetes-csi/external-snapshotter/pull/7.  Close this one.

**Pull request closed without merging**